mu/virtual_environment.py

```diff
@@ -478,24 +516,52 @@ def ensure_and_create(self, emitter=None):
 
         n_tries = 2
         n = 1
-        #
-        # First time around we'll have a path with nothing in it
-        # Jump straight to creation and let the ensure take over afterwards
-        # For the happy path we'll get a clean creation and a clean ensure
-        #
         try_to_create = not os.path.exists(self.path)
         while True:
             logger.debug("Checking virtual environment; attempt #%d.", n)
             try:
+                #
+                # First time around we'll have a path with nothing in it
+                # Jump straight to creation and let the ensure take over afterwards
+                # For the happy path we'll get a clean creation and a clean ensure
+                #
                 if try_to_create:
                     self.create()
+                    try_to_create = False
+
+                #
+                # Otherwise check whether the venv settings match the version of Mu
+                # If not, recreate the venv
+                #
+                else:
+                    venv_mu_version = self.settings.get(
+                        "mu_version", "-no-version-"
+                    )
+                    if venv_mu_version != mu_version:
+                        logger.warning(
+                            "Venv created by Mu version %s; Current Mu is version %s",
+                            venv_mu_version,
+                            mu_version,
+                        )
+                        self.recreate()

```

Comment:
I guess this means every time we use a different Mu version it will recreate the venv? Even if venv was "compatible".
Would it be good enough to `self.ensure()` first before recreating if the version is different?

Reply:
Well that's basically what we're doing right now, and we then run into the problems of having two slightly different versions of whatever. I started off trying pick things apart -- uninstalling older versions of baseline packages etc. -- but I came to the conclusion that creating a new venv and porting across the user packages was a safe option.

Reply:
> Well that's basically what we're doing right now, and we then run into the problems of having two slightly different versions of whatever. 

Right, yes, but that's only for the cases where upgrading has incompatibilities. We probably had quite a few people upgrading Mu from beta.2 to beta.3 without having to recreate the venv, and without issues.
The thing that makes me a bit nervous is that sooner or latter at the point where we "port" the user packages to  a new venv, then something will fail due to the pip resolver finding incompatibilities. So the longer we can reuse a venv the better, the idea being that if something is not compatible `self.ensure()` should fail anyway? 

> I started off trying pick things apart -- uninstalling older versions of baseline packages etc. -- but I came to the conclusion that creating a new venv and porting across the user packages was a safe option.

Oh yeah, 100% agree.

Reply:
I think for this release I'm going to stay with the over-cautious approach; otherwise we'll never get the release out. I entirely agree that there's a risk in the other direction: that we might "damage" a perfectly good venv. That said, we have a fallback position (albeit a manual one) because we don't remove the existing venv and we can always update `venv.json` to point to it.